File: sqrt.py
#Approach 1- Binary Search
def mySqrt(x):          #Time complexity- O(log x) (Binary search tree)
    if x==0 or x==1:
        return x
    start=1
    end=x//2
    while start<=end:
        mid=start+(end-start)//2 #To avoid overflow we use start+(end-start)//2 instead of simple (start+end)//2
        if mid > x/mid:      # which is same as mid*mid>x:                          
            end=mid-1
        else:
            start=mid+1
            
    return start-1

#Approach 2- Recursion + Bit shifts 
def mySqrt1(x):
    if x < 2:
        return x

    left=mySqrt(x>>2)<<1
    right=left+1

    return left if right*right > x else right 

#Approach 3-  Newton's method (It is the fastest method)
def mySqrt2(x): # TC O(log n)
    if x < 2:
        return x 

    x0 = x
    x1 = (x0 + x / x0) / 2
    while abs(x0 - x1) >= 1:
        x0 = x1
        x1 = (x0 + x / x0) / 2        

    return int(x1)

print(mySqrt(16))
print(mySqrt1(49))
print(mySqrt2(729))

File: sqrt.java
class Solution{
	public int sqrt(int n){
		int left=0;
		int right=n/2;
		while (left<=right){
			int mid= (left+right)/2;
			long num=(long)mid*mid;
			if (num<n){
				left=mid+1;
			}else if (num>n){
				right=mid-1;
			}
			else{
				return mid;
			}
		}
		return right;
	}
}

class Solution1{
	public int sqrt(int n){
		if (n < 2) return n;

		int left=sqrt(n>>2)<<1;
		int right=left+1;
		return (long)right*right>n? left:right;

	}
}

class Solution2{
	public int sqrt(int n){
		if (n<2) return n;
		
		double x0=n;
		double x1=(x0+n/x0)/2.0;
		while (Math.abs(x0-x1)>=1){
			x0=x1;
			x1=(x0+n/x0)/2.0;
		}
		return (int)x1;
	}
}

class sqrt{
	public static void main(String[] args){
		int n=529; 
		Solution s= new Solution();
		System.out.println(s.sqrt(n));
		Solution1 s1= new Solution1();
		System.out.println(s1.sqrt(n));
		Solution2 s2= new Solution2();
		System.out.println(s2.sqrt(n));
	}
}